File: leetcode/SearchforaRange/searchRange.java
public class Solution {
    public int[] searchRange(int[] A, int target) {
        // Start typing your Java solution below
        // DO NOT write main() function
        int ans[] = new int[2];
        int left = bs(A, target - 1), right = bs(A, target);
        if (right != -1 && A[right] != target) right = -1;
        ans[1] = right;
        ans[0] = right == -1 ? -1 : left + 1;
        return ans;
    }
    private int bs(int[] A, int target) {
        int len = A.length, low = 0, high = len - 1, mid;
        while (low <= high) {
            mid = low + (high - low) / 2;
            if (A[mid] > target) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return high;
    }
}

File: leetcode/SearchforaRange/searchRange.cpp
class Solution {
public:
    int helper(vector<int> &nums, int target) {
        // find the start position of number equal or larger than target
        int l = 0, r = nums.size();
        while (l < r) {
            int m = l + (r - l) / 2;
            if (nums[m] < target) {
                l = m + 1;
            } else r = m;
        }
        return l;
    }
    vector<int> searchRange(vector<int>& nums, int target) {
        int left = helper(nums, target);
        if (left == nums.size() || nums[left] != target) return {-1, -1};
        return {left, helper(nums, target + 1) - 1};
    }
};
